### ISSUE TITLE
Bitbucket IP range check not working

### ISSUE THREAD

**Issue by nlsrchtr** [adnanh/webhook#135]
Thanks for providing this very useful tool!

I trying to use a webhook from Bitbucket, as explained on the [examples](https://github.com/adnanh/webhook/wiki/Hook-Examples#incoming-bitbucket-webhook). But using the IP range for whitelisting, the response is "Hook rules were not satisfied.". Removing the whitelist restriction, everything works fine.

I was expecting, that running `webhook --hooks /etc/webhook.conf -verbose` to see the IP address of the requesting servers, but this is not the case. Since Bitbucket [also offers IPV6 IP addresses](https://confluence.atlassian.com/bitbucket/what-are-the-bitbucket-cloud-ip-addresses-i-should-use-to-configure-my-corporate-firewall-343343385.html), it maybe the case, that they maybe used.

Is there any way to determine the IP address of the requesting servers? Is IPV6 supported?

Thanks for your feedback.

**Comment by adnanh** [comment]
Hey, thanks for using webhook :-)

I will add IP logging support in the next release. I agree that it's a valuable piece of information that you currently cannot see in the logs. Currently, the only way to check for the incoming IP is to set up a proxy in front of webhook that would pass it down via the header, or, shut down webhook and start up a netcat on the port webhook is listening on and see the ip :-)

**Comment by nlsrchtr** [comment]
@adnanh Thanks for your feedback and for adding IP logging in future releases. I'll dig into netcat. Thanks for the idea.

**Comment by nlsrchtr** [comment]
I looked into it with `nc -l 9000 -v`. BitBucket sent requests with `104.192.143.193` and `104.192.143.194`. I'm using the exact sample piece from the example:

```json
"trigger-rule":
{
  "match":
  {
    "type": "ip-whitelist",
    "ip-range": "104.192.143.0/24"
  }
}
```
The response I see at BitBucket still is `Hook rules were not satisfied.` - since this is the only rule, I'm wondering why this doesn't match and work?

Maybe you have a chance to try it on your side?

**Comment by adnanh** [comment]
I'll try it out as soon as I get some free time :-)

**Comment by adnanh** [comment]
It appears to be working for me. Could you please post the full hooks json file?

**Comment by adnanh** [comment]
Also, I've released 2.6.4 with the IP logging feature

**Comment by nlsrchtr** [comment]
@adnanh Thanks for taking the timing! My configuration looks like:

```json
[
  {
    "id": "puppetmaster-webhook",
    "execute-command": "PATH_TO_EXECUTABLE",
    "command-working-directory": "PATH_TO_WORKING_DIRECTORY",
    "include-command-output-in-response": true,
    "trigger-rule":
    {
      "match":
      {
        "type": "ip-whitelist",
        "ip-range": "104.192.143.0/24"
      }
    }
  }
]
```

So it's just the example taken from the "Hook examples". The response at Bitbucket displays "Hook rules were not satisfied." as body content.

Since I use Ubuntu packaging to install and run webhook, could you give me short instruction howto update to the latest version? Since it's Go bases, I could just download the latest executable, right?

**Comment by adnanh** [comment]
Yes, depending on your ARCH, all you need is to download the appropriate webhook-linux-ARCH.tar.gz from the Releases tab. (Most likely it'll be this one: https://github.com/adnanh/webhook/releases/download/2.6.4/webhook-linux-amd64.tar.gz)

**Comment by nlsrchtr** [comment]
Thanks for taking the time investigating this in more. It works for me with the new release, but not with the old, with the same config:

**New version, without IP whitelisting in config**
```
[webhook] 2017/06/08 14:59:46 incoming HTTP request from 104.192.143.193:61525
[webhook] 2017/06/08 14:59:46 XYZ-webhook got matched
[webhook] 2017/06/08 14:59:46 XYZ-webhook hook triggered successfully
[webhook] 2017/06/08 14:59:46 executing /... (/...) with arguments ["/.."] and environment [] using /etc/.../code/ as cwd
[webhook] 2017/06/08 14:59:48 command output:
[webhook] 2017/06/08 14:59:48 finished handling XYZ-webhook
[webhook] 2017/06/08 14:59:48 2017-06-08T14:59:46Z | 200 | 	 1.591855914s | SERVER:9000 | POST /hooks/XYZ-webhook
```
Here I changed the config to enable IP whitelisting.

**New version, with IP whitelisting in config**
```
[webhook] 2017/06/08 15:00:46 incoming HTTP request from 104.192.143.194:63008
[webhook] 2017/06/08 15:00:46 XYZ-webhook got matched
[webhook] 2017/06/08 15:00:46 XYZ-webhook hook triggered successfully
[webhook] 2017/06/08 15:00:46 executing /... (/...) with arguments ["...t"] and environment [] using /etc/.../code/ as cwd
[webhook] 2017/06/08 15:00:47 command output:
[webhook] 2017/06/08 15:00:47 finished handling XYZ-webhook
[webhook] 2017/06/08 15:00:47 2017-06-08T15:00:46Z | 200 | 	 1.739314364s | SERVER:9000 | POST /hooks/XYZ-webhook
```
**Back to old version, with IP whitelist enabled**
```
[webhook] 2017/06/08 15:01:51 Started POST /hooks/XYZ-webhook
[webhook] 2017/06/08 15:01:51 XYZ-webhook got matched
[webhook] 2017/06/08 15:01:51 XYZ-webhook got matched, but didn't get triggered because the trigger rules were not satisfied
[webhook] 2017/06/08 15:01:51 Completed 200 OK in 387.294µs
```
**Back to new version, with IP whitelist enabled, with invalid IP**
```
[webhook] 2017/06/08 15:06:56 incoming HTTP request from 213.61.242.154:52148
[webhook] 2017/06/08 15:06:56 XYZ-webhook got matched
[webhook] 2017/06/08 15:06:56 XYZ-webhook got matched, but didn't get triggered because the trigger rules were not satisfied
[webhook] 2017/06/08 15:06:56 2017-06-08T15:06:56Z | 200 | 	 127.205µs | SERVER:9000 | POST /hooks/XYZ-webhook
```

The problem seems to be solved for me, but I'm wondering if you changed more than just adding logging of the IP address?

Thanks again for your time!

_I anonymized paths and server name_

**Comment by adnanh** [comment]
Nope, I literally fixed tests that were failing due to bad fixture. Here are the 2.6.4 changes: https://github.com/adnanh/webhook/commit/a38875213a92171fd9779629f77375691ee2b9c9

**Comment by adnanh** [comment]
Does the trigger rule fail in the new version when you hit it from an IP that is not whitelisted? Did you compile both versions from source? I might have upgraded Golang version in between them on my laptop...

**Comment by adnanh** [comment]
I'll close this issue for now as it appears to be fixed in 2.6.4. Feel free to reopen if it resurfaces again...
